README.md
# hsms

A simple Hardware Security Module ([HSM](https://en.wikipedia.org/wiki/Hardware_security_module)) simulator. 
The list of supported commands is:

Only the double-length key 3DES operations are supported.


## To use:
```javascript
const hsm = new HSM()

```

## To customize:

```bash
git clone https://github.com/timgabets/hsms.git
cd hsms
# Install all the dependencies
npm install
# Init ava test framework
ava --init
# Check the tests
npm test
```


CHANGELOG
October 2017
	0.1.0	Initial commit

tests/hsm.js
import test from 'ava';
import HSM from '../hsm'

var test_lmk = 'DEAFBEEDEAFBEEDEAFBEEDEAFBEEDEAF'
var hsm = new HSM(test_lmk)

test('should assign LMK', t => {
  t.is(hsm.LMK, test_lmk)
});

test('should get 6-digit LMK check value by default', t => {
  t.is(hsm.getKeyCheckValue(test_lmk), 'F4EDC8')
});

test('should get arbitrary-length LMK check value', t => {
  t.is(hsm.getKeyCheckValue(test_lmk, 16), 'F4EDC8DEB67F6E28')
});

test('should get diagnostics data', t => {
  var response = hsm.getDiagnosticsData()

  t.is(response.get('Response Code'), 'ND')
  t.is(response.get('Error Code'), '00')
  t.is(response.get('LMK Check Value'), 'F4EDC8DEB67F6E28')
  t.is(response.get('Firmware Version'), '0007-E000')
});

test('should be able to get proper empty message length', t => {
  t.is(hsm.getOutgoingMessageLength(''), '\x00\x00')  
});

test('should return the length for message.len === 10', t => {
  t.is(hsm.getOutgoingMessageLength('0123456789'), '\x00\x0a')  
});

test('should return the length for message.len === 16', t => {
  t.is(hsm.getOutgoingMessageLength('0123456789abcdef'), '\x00\x10')  
});

test('should return the length for message.len === 255', t => {
  t.is(hsm.getOutgoingMessageLength('123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef'), '\x00\xff')  
});

test('should return the length for message.len === 256', t => {
  t.is(hsm.getOutgoingMessageLength('0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef'), '\x01\x00')  
});

test('should return the length for message.len === 2636', t => {
  t.is(hsm.getOutgoingMessageLength('0123456789abcdef012340123456789abcdef01234567889abcdef01289abcdef01289abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01234567893456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01234567893456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01234567899abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef567889abcdef01289abcdef01289abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01234567893456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01234567893456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01234567899abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef'), '\x0a\x4c')  
});

test('should build data string from response object', t => {
  var response = new Map()
    
  response.set('Response Code', 'ND');
  response.set('Error Code','00');
  response.set('LMK Check Value', 'F4EDC8DEB67F6E28');
  response.set('Firmware Version', '0007-E000');

  t.is(hsm.build(response), '\x00\x1dND00F4EDC8DEB67F6E280007-E000')
});

test('should build data string from response object with header', t => {
  var response = new Map()
    
  response.set('Response Code', 'ND');
  response.set('Error Code','00');
  response.set('LMK Check Value', 'F4EDC8DEB67F6E28');
  response.set('Firmware Version', '0007-E000');

  hsm.header = 'SSSS'

  t.is(hsm.build(response), '\x00\x21SSSSND00F4EDC8DEB67F6E280007-E000')
});

hsm.js
import {des as des3} from 'node-cardcrypto'

class HSM{
  constructor(lmk, header) {
    lmk ? this.LMK = lmk : this.LMK = 'DEAFBEEDEAFBEEDEAFBEEDEAFBEEDEAF'
    header ? this.header = header : this.header = ''

    this.firmware_version = '0007-E000';
  }

  test(){
    return true
  }

  /**
   * [getKeyCheckValue description]
   * @param  {[type]} key [description]
   * @return {[type]}     [description]
   */
  getKeyCheckValue(key, cv_length){
    var kcv = des3.ecb_encrypt(key, '00000000000000000000000000000000');
    if(kcv){
      if(cv_length)
        return kcv.substr(0, cv_length);
      else
        return kcv.substr(0, 6);
    }
    else
      return null;
  }

  /**
   * [getDiagnosticsData description]
   * @return {[type]} [description]
   */
  getDiagnosticsData(){
    var response = new Map()
    
    response.set('Response Code', 'ND');
    response.set('Error Code','00');
    response.set('LMK Check Value', this.getKeyCheckValue(this.LMK, 16));
    response.set('Firmware Version', this.firmware_version);

    return response;
  }

  getResponse(data){
    switch(data.substr(0,2)){
      case 'NC':
        return this.getDiagnosticsData()
    }
  }

  /**
   * [getOutgoingMessageLength description]
   * @param  {[type]} data [description]
   * @return {[type]}      [description]
   */
  getOutgoingMessageLength(data){
    return String.fromCharCode(data.length / 256) + String.fromCharCode(data.length % 256);
  };

  /**
   * [build description]
   * @param  {[type]} response [description]
   * @return {[type]}          [description]
   */
  build(response){
    var message = this.header;
    for (var [key, value] of response)
      message += value

    return this.getOutgoingMessageLength(message) + message
  }
}

module.exports = HSM
